### ISSUE TITLE
Memory leak in Mutiny with Quarkus integration

### ISSUE THREAD

**Issue by Settis** [quarkusio/quarkus#38654]
### Describe the bug

We serve request via Uni, by attaching callbacks for processing the item and providing timeouts. The callbacks links with state we need to process the request. It's expected that after the Uni will be done, GC can collect this big related state. In fact the state is kept in memory untill the timeout.

### Expected behavior

All callback-related objects can be collected by GC because they are not needed for completed Uni.

### Actual behavior

The related objects are kept until the someTimout is passed.

### How to Reproduce?

- Create a Uni, and put here some callbacks.
- On top of that call .ifNoItem().after(someTimeout).fail()
- Complete the Uni.

Example: https://github.com/Settis/mutinyInQuarkus

### Output of `uname -a` or `ver`

Linux agrigorev 6.7.3-arch1-2 #1 SMP PREEMPT_DYNAMIC Fri, 02 Feb 2024 17:03:55 +0000 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.10" 2024-01-16 LTS OpenJDK Runtime Environment Zulu17.48+15-CA (build 17.0.10+7-LTS) OpenJDK 64-Bit Server VM Zulu17.48+15-CA (build 17.0.10+7-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.2.7.Final and 3.7.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.7 (b89d5959fcde851dcb1c8946a785a163f14e1e29) Maven home: /opt/maven Java version: 17.0.10, vendor: Azul Systems, Inc., runtime: /usr/lib/jvm/zulu-17 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "6.7.3-arch1-2", arch: "amd64", family: "unix"

**Comment by jponge** [comment]
1. Can you please update to the latest Quarkus version?
2. It is not clear to me what you call a memory leak, even looking at https://github.com/Settis/mutinyInQuarkus/blob/master/src/test/java/org/acme/UniLeakTest.java

**Comment by jponge** [comment]
I missed the 3.7.1 part in:

> 3.2.7.Final and 3.7.1

Can you please offer a more realistic reproducer? What you have here doesn't really help / some objects can't be GC'ed by design until after the `Uni` pipeline has completed, and the timeout expired (worst case).

**Comment by Settis** [comment]
The main idea here is that those objects are GC'ed in pure Mutiny after the Uni processed. But in Quarkus they are stiked inside `EnhancedQueueExecutor` untill the timeout.
And if you have relatively big timeout you have a chanse to collect a lot data related to already processed requests.

**Comment by jponge** [comment]
That's interesting

**Comment by jponge** [comment]
@Settis Thanks! Also do you confirm that this doesn't happen in plain Mutiny when running the test outside of Quarkus integration?

**Comment by Settis** [comment]
That's right, it does not happen in plain Mutiny. Only with the `EnhancedQueueExecutor`.

Other observations:
- inside `Infrastructure::setDefaultExecutor` the executor may be wrapped by `MutinyScheduler`. It is used in both cases with plain Mutiny and Quarkus integration, so it seems not related to the issue.
- inside Quarkus test I can first run `Infrastructure.setDefaultExecutor()` that thing changes Mutiny executor from `EnhancedQueueExecutor` back to default one. And GC'ing work as expected. Unfortunately, before changing executors Mutiny shut down the current one, and this broke the whole Quarkus, so I can't use this as a fix :)

**Comment by Settis** [comment]
For information if someone else found this thread.

The `Uni` result is kept too until the timeout. As a workaround, you can schedule `UniOnTimeout` in another scheduler like this: `.ifNoItem().after(Duration.ofSeconds(10)).on(yourService)`.
